### ISSUE TITLE
Make the Dogelog runtime variables pre trailed

### ISSUE THREAD

**Issue by jburse** [jburse/dogelog-moon#153]
We could make Dogelog runtime variables pre trailed. We made the
Jekejeke Prolog variables already pre trailed. This means they
contain fields for the trail container datastructure, and it is

not that to add the variable trail container we would need the
creation of some extra object. In the case of Jekejeke Prolog it
was a double linked list. In the case of Dogelog runtime we

only need a single linked list. Subsequently we wouldn't need the
List datastructure anymore and the change also demands to rewrite
the part of the garbage collector that adjust or sweep the trail.

**Comment by jburse** [comment]
We already implemented the change for the JavaScript version.
When we were at it, we also helped the host language garbage
collector a little bit, by undoing variable binding in the Dogelog

runtime sweep phase. This instantiation was not set to undefined.
We now set the instantiantion to undefined and the tail of the
pre-trail to null. Need to do the Python version as well.

**Comment by jburse** [comment]
We sticked to a solution that also implements the new feature
that it agressively nulls Prolog garbage collected variables. The advantage
of such an approach is that it finds programming errors.

We found for example that we forgot a flip_color() call in the
Python version. The Python test case now runs slower, because it
does now more correctly do its full Prolog garbage collection:
```
/* Before, Buggy */
% ?- time(fibo(30, X)).
% % Wall 136649 ms, trim 2005 ms
% X = 1346269.

/* After, Corrected */
?- time(fibo(30,X)).
% Wall 140254 ms, trim 1870 ms
X = 1346269.
```
On the other hand we see a 10% speed increase for the JavaScript
version. Running with node.js from the Windows command line,
we see the following improvement:
```
/* Before */
?- time(fibo(30, X)).
% Wall 4590 ms, trim 41 ms
X = 1346269.

/* After */
?- time(fibo(30, X)).
% Wall 4108 ms, trim 38 ms
X = 1346269.
```
All measurements are warm runs. After successful implementation,
closing the ticket now.
